### PR TITLE
Fix panic with 'stuck spinlock' during crash recovery tests

### PR DESCRIPTION
**Pull request by whitehawk**
Fix panic with 'stuck spinlock' during crash recovery tests

Problem:
Isolation crash recovery tests (for example, uao_crash_compaction_column, 
crash_recovery_dtm, etc.) failed with 'stuck spinlock' panic with very low
reproduction rate.

Cause:
Problem happened in the following case:
1. During the test, a backend process was forced into a panic (according to test
scenario).
2. At the very same moment, the background writer process called 
'SIMPLE_FAULT_INJECTOR("fault_in_background_writer_main")', he did it regularly
from the loop in 'BackgroundWriterMain'. The bg writer acquired the spinlock
inside the fault injector, but before it released it...
3. Postmaster started to send the SIGQUIT signal to all child processes.
4. Background writer process received SIGQUIT and halted its execution without
releasing the spinlock.
5. In the background writer process, a handler for the SIGQUIT signal was 
invoked - 'bg_quickdie'. And 'bg_quickdie' called 
'SIMPLE_FAULT_INJECTOR("fault_in_background_writer_quickdie");'. As the spinlock
was not released, 'bg_quickdie' hanged on the spinlock.
6. Postmaster failed to wait child processes completion. And after timeout it
tried to check 'SIMPLE_FAULT_INJECTOR("postmaster_server_loop_no_sigkill")' and
hanged on the same spinlock as well.
7. Finally, both postmaster and 'bg_quickdie' failed with 'stuck spinlock'
panic.

Fix:
The general rule would be that it is NOT allowed to access the fault injector 
from the postmaster process or from any SIGQUIT handler of child processes when 
the system is resetting after the crash of some backend. It is so because during 
reset postmaster terminates child processes, and it might terminate some process 
when the process has acquired the spinlock of the fault injector, but hasn't 
released it yet. So, subsequent calls to the fault injector api from postmaster 
or any SIGQUIT handler will lead to deadlock. Only 'doomed' processes can still 
call the fault injector api, as they will soon be terminated anyway. According 
to written above, following changes are made:
1. Access to the fault injector during a reset is removed from 'bg_quickdie' and 
postmaster's ServerLoop. 
2. The fts_segment_reset test and related code are redesigned. Now, instead of 
sleeping for a delay defined by the test in 'bg_quickdie', postmaster sends 
SIGSTOP to the bg writer, and starts a timer for the delay. Once the delay 
elapses, the timer handler sends SIGCONT and SIGQUIT to the bg writer. This 
logic is triggered if the postmaster detects the new fault 
"postmaster_delay_termination_bg_writer" (which is a replacement for 
"fault_in_background_writer_quickdie" and "postmaster_server_loop_no_sigkill"). 
This fault is checked only when the postmaster is in the PM_RUN state. So it
should be safe to check for it. 

New tests specific to this issue are not added because of the unstable nature of
the problem.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67342

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200047

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67375

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200320

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200321

File: src/backend/postmaster/postmaster.c [hunk 1]
```diff
@@ -1963,6 +1970,39 @@ ServerLoop(void)
...
+			 * and it might kill some process when the process has acquired
+			 * the spin lock of the fault injector, but hasn't released it
+			 * yet. So, subsequent call of the fault injector api from
+			 * Postmaster or any SIGQUIT handler will lead to deadlock. Only
+			 * 'doomed' processes can still call the fault injector api, as
+			 * they will soon be terminated anyway. Thus, the fault injector
+			 * is accessed only when the Postmaster state is PM_RUN.
+			 */
+			if (SIMPLE_FAULT_INJECTOR_WITH_PARAM("postmaster_delay_termination_bg_writer", \
+										 &delay_bg_writer_termination_sec) ==
```
Comment (bandetto) [inline src/backend/postmaster/postmaster.c]: Maybe this indentation will be better:
```suggestion
			if (SIMPLE_FAULT_INJECTOR_WITH_PARAM("postmaster_delay_termination_bg_writer",
												 &delay_bg_writer_termination_sec) == FaultInjectorTypeSkip)
```

File: src/backend/postmaster/postmaster.c [hunk 2]
```diff
@@ -3873,11 +3912,37 @@ HandleChildCrash(int pid, int exitstatus, const char *procname)
...
+			 * In order to delay the termination of the background writer
+			 * process - send SIGSTOP to the bg writer process and start a
+			 * timer that will wake up and then immediately terminate it.
+			 */
+			ereport(DEBUG2,
+					(errmsg_internal("sending SIGSTOP to process %d and starting timer",
+									 (int) BgWriterPID)));
+			signal_child(BgWriterPID, SIGSTOP);
+
+			static TimeoutId id = 0;
```
Comment (bandetto) [inline src/backend/postmaster/postmaster.c]: Definition should be moved to the start of the block. Why does it need to be static?

File: src/include/utils/faultinjector.h [hunk 3]
```diff
@@ -115,9 +121,13 @@ extern void HandleFaultMessage(const char* msg);
 extern bool am_faulthandler;
 #define SIMPLE_FAULT_INJECTOR(FaultName) \
 	FaultInjector_InjectFaultIfSet(FaultName, DDLNotSpecified, "", "")
+
+#define SIMPLE_FAULT_INJECTOR_WITH_PARAM(FaultName, ParamPtr) \
+	FaultInjector_InjectFaultIfSetWithParam(FaultName, DDLNotSpecified, "", "", ParamPtr)
```
Comment (bandetto) [inline src/include/utils/faultinjector.h]: Do we really need a an extra parameter? If we didn't need such function before, and right now we don't use values other than 17 for the test, can't the value for `postmaster_delay_termination_bg_writer` fault just be hard-coded in?

File: src/backend/postmaster/postmaster.c [hunk 4]
```diff
@@ -155,6 +155,9 @@ bool am_mirror = false;
 /* GPDB specific flag to handle deadlocks during parallel segment start. */
 volatile bool *pm_launch_walreceiver = NULL;
 
+/* This is set by a test to simulate long RESET state. */
+static int	delay_bg_writer_termination_sec = 0;
```
Comment (bandetto) [inline src/backend/postmaster/postmaster.c:159]: Maybe `postmaster_no_sigkill` should be defined here as well, and put behind a `#ifdef FAULT_INJECTOR`?

File: src/backend/postmaster/postmaster.c [hunk 5]
```diff
@@ -5676,6 +5741,21 @@ dummy_handler(SIGNAL_ARGS)
 {
 }
 
+/*
+ * Terminate stopped background writer process (should be invoked only
+ * during test).
+ */
+static void
+TerminateBgWriterHandler()
```
Comment (bandetto) [inline src/backend/postmaster/postmaster.c:5756]: Maybe this should be but behind `#ifdef FAULT_INJECTOR`?

File: src/backend/postmaster/postmaster.c [hunk 6]
```diff
@@ -738,6 +742,8 @@ PostmasterMain(int argc, char *argv[])
 	 */
 	InitializeGUCOptions();
 
+	InitializeTimeouts();
```
Comment (bandetto) [inline src/backend/postmaster/postmaster.c:745]: Seems like `InitializeTimeouts()` [resets all previously active timeouts](https://github.com/arenadata/gpdb/blob/46fb51ca8e4dea8602fca7ffa23b3cb7110b3b33/src/backend/utils/misc/timeout.c#L389-L397). Is this call safe?

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67528

File: src/backend/postmaster/postmaster.c (hunk 1 above)
Comment (whitehawk) [reply src/backend/postmaster/postmaster.c]: > Maybe this indentation will be better:

Updated, but put the indentation provided by 'pgindent'

File: src/backend/postmaster/postmaster.c (hunk 2 above)
Comment (whitehawk) [reply src/backend/postmaster/postmaster.c]: > Definition should be moved to the start of the block.
 
Yes, moved it.

> Why does it need to be static?

It has to be static, because we need to call `RegisterTimeout` only once before first call of `enable_timeout_after`.
And on subsequent calls we reuse the timer id. If `static` is removed, soon we will run out of available timers.

File: src/include/utils/faultinjector.h (hunk 3 above)
Comment (whitehawk) [reply src/include/utils/faultinjector.h]: > Do we really need a an extra parameter? If we didn't need such function before, and right now we don't use values other than 17 for the test, can't the value for postmaster_delay_termination_bg_writer fault just be hard-coded in?

Value of `17` is something chosen by the logic of the test. Better to keep coupled things together. 

I agree that hard coding the value will allow to reduce the delta. But the hard-coded value here will be very unnatural.
Every time one facing this piece of code will need to refer the test to understand the logic behind.

If you insist on this change, please involve 2nd reviewer for more opinions.

File: src/backend/postmaster/postmaster.c (hunk 4 above)
Comment (whitehawk) [reply src/backend/postmaster/postmaster.c:159]: > Maybe postmaster_no_sigkill should be defined here as well, and put behind a #ifdef FAULT_INJECTOR?

If it was just one single piece of code to wrap, I would agree.
But in this patch there are several places changed, and wrapping every of them significantly reduces code readability .
It will lead to code like below:
```
#ifdef FAULT_INJECTOR
		if (delay_bg_writer_termination_sec == 0)
#endif
		{
			ereport(DEBUG2,
					(errmsg_internal("sending %s to process %d",
									 (SendStop ? "SIGSTOP" : "SIGQUIT"),
									 (int) BgWriterPID)));
			signal_child(BgWriterPID, (SendStop ? SIGSTOP : SIGQUIT));
		}
#ifdef FAULT_INJECTOR
		else
		{
			/*
			 * Code below is only for test purposes and should not be invoked
			 * under normal conditions. It will delay the termination of the
			 * background writer process.
			 *
			 * In order to delay the termination of the background writer
			 * process - send SIGSTOP to the bg writer process and start a
			 * timer that will wake up and then immediately terminate it.
			 */
			static TimeoutId id = 0;

			ereport(DEBUG2,
					(errmsg_internal("sending SIGSTOP to process %d and starting timer",
									 (int) BgWriterPID)));
			signal_child(BgWriterPID, SIGSTOP);

			if (id == 0)
				id = RegisterTimeout(USER_TIMEOUT, TerminateBgWriterHandler);

			enable_timeout_after(id, delay_bg_writer_termination_sec * 1000);
		}
#endif
```
I'd like to avoid such style if possible. Readability is important.

File: src/backend/postmaster/postmaster.c (hunk 5 above)
Comment (whitehawk) [reply src/backend/postmaster/postmaster.c:5756]: > Maybe this should be but behind #ifdef FAULT_INJECTOR?

The same comment as above.

File: src/backend/postmaster/postmaster.c (hunk 6 above)
Comment (whitehawk) [reply src/backend/postmaster/postmaster.c:745]: > Seems like InitializeTimeouts() resets all previously active timeouts. Is this call safe?

Postmaster didn't initialize timers before my change (otherwise I wouldn't need this change).
And this place is before start of other processes (though as far as I see it shouldn't interfere with the timers of other processes).
So yes, I think it is safe to call it here.

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1209547

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1209548

File: src/backend/postmaster/postmaster.c [hunk 7]
```diff
@@ -1963,6 +1970,38 @@ ServerLoop(void)
...
+			 * 'doomed' processes can still call the fault injector api, as
+			 * they will soon be terminated anyway. Thus, the fault injector
+			 * is accessed only when the Postmaster state is PM_RUN.
+			 */
+			if (SIMPLE_FAULT_INJECTOR_WITH_PARAM("postmaster_delay_termination_bg_writer",
+				  &delay_bg_writer_termination_sec) == FaultInjectorTypeSkip)
+				postmaster_no_sigkill = true;
+			else
+			{
+
```
Comment (bandetto) [inline src/backend/postmaster/postmaster.c]: Empty line

File: src/include/utils/faultinjector.h (hunk 3 above)
Comment (bandetto) [reply src/include/utils/faultinjector.h]: If extra parameter would be removed, think the fault name can be changed to something like `postmaster_delay_termination_bg_writer_by_17s` to give some clues. I will leave this question for the second reviewer.

**Review by bandetto — COMMENTED** [review]
Given we stop the Postmaster from killing children with `SIMPLE_FAULT_INJECTOR("postmaster_server_loop_no_sigkill")`, why are we encountering the deadlock? Can you explain this point in the description?

**Comment by whitehawk** [comment]
> Given we stop the Postmaster from killing children with `SIMPLE_FAULT_INJECTOR("postmaster_server_loop_no_sigkill")`, why are we encountering the deadlock? Can you explain this point in the description?

Refer to the comments just above the place where the `SIMPLE_FAULT_INJECTOR("postmaster_server_loop_no_sigkill")` was:
```
		/*
		 * If we already sent SIGQUIT to children and they are slow to shut
		 * down, it's time to send them SIGKILL.  This doesn't happen
		 * normally, but under certain conditions backends can get stuck while
		 * shutting down.  This is a last measure to get them unwedged.
		 *
		 * Note we also do this during recovery from a process crash.
		 */
```

So by this time all child processes should be terminated (refer to `reaper`). If someone hangs, postmaster killd it with SIGKILL from here.
But that is something we do not want for the test. Thus "postmaster_server_loop_no_sigkill" prevented sending SIGKILL to the bg writer during the test.

File: src/backend/postmaster/postmaster.c (hunk 7 above)
Comment (whitehawk) [reply src/backend/postmaster/postmaster.c]: fixed

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67581

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1211760

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1211761

**Comment by bandetto** [comment]
> So by this time all child processes should be terminated (refer to reaper). If someone hangs, postmaster killd it with SIGKILL from here.

So the main point is that the Postmaster's SIGQUIT can be received while Fault Injector's lock is still taken, leaving it locked forever. But in the description, it is stated that Postmaster "kills", which is slightly confusing. The PR description should be reworded to clarify how SIGQUIT and SIGKILL work together in this context. 

Additionally, the "Reason: " in the "Fix: " paragraph should be a separate title, judging by the other titles. The "Fix: " and "Solution is following: " sections can be merged to make description more concise. In "Solution is following: ", points 1, 2 may be omitted, due to being easy to find in the PR diff.

When reading the description, it should be easy to understand which behavior was incorrect and has been corrected, and which behavior is correct and remains unchanged. One way to differentiate this could be using past and present tense to indicate this distinction:
may fail with 'stuck spinlock' panic --> would fail with 'stuck spinlock' panic;
'SIMPLE_FAULT_INJECTOR_WITH_PARAM' was added --> 'SIMPLE_FAULT_INJECTOR_WITH_PARAM' is added;

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1211755

**Comment by whitehawk** [comment]
> So the main point is that the Postmaster's SIGQUIT can be received while Fault Injector's lock is still taken, leaving it locked forever. But in the description, it is stated that Postmaster "kills", which is slightly confusing. The PR description should be reworded to clarify how SIGQUIT and SIGKILL work together in this context.
> 
> Additionally, the "Reason: " in the "Fix: " paragraph should be a separate title, judging by the other titles. The "Fix: " and "Solution is following: " sections can be merged to make description more concise. In "Solution is following: ", points 1, 2 may be omitted, due to being easy to find in the PR diff.
> 
> When reading the description, it should be easy to understand which behavior was incorrect and has been corrected, and which behavior is correct and remains unchanged. One way to differentiate this could be using past and present tense to indicate this distinction: may fail with 'stuck spinlock' panic --> would fail with 'stuck spinlock' panic; 'SIMPLE_FAULT_INJECTOR_WITH_PARAM' was added --> 'SIMPLE_FAULT_INJECTOR_WITH_PARAM' is added;

Updated the description, please check

File: src/include/utils/faultinjector.h (hunk 3 above)
Comment (bandetto) [reply src/include/utils/faultinjector.h]: Regarding the last meeting, please let me know about any updates regarding the possibility of removing the strict 17 second timeout.

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67764

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1220815

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1220816

File: src/test/isolation2/sql/fts_segment_reset.sql [hunk 8]
```diff
@@ -12,15 +12,15 @@
 -- to restart, and potentially makes FTS think it's in "recovery not
 -- in progress" stage and promote the mirror, we would need the FTS
 -- to make that decision a bit less frequently.
-!\retcode gpconfig -c gp_fts_probe_retries -v 15 --masteronly;
+!\retcode gpconfig -c gp_fts_probe_retries -v 15;
```
Comment (bandetto) [inline src/test/isolation2/sql/fts_segment_reset.sql:15]: Why is this change needed? FTS runs on the master only.

Comment (whitehawk) [reply src/test/isolation2/sql/fts_segment_reset.sql:15]: Because the crash will be handled by segment, and we need to know `gp_fts_probe_retries` on the segment to set properly `delay_bg_writer_termination_sec`. So we need to propagate this  guc to the segments as well.

**Review by bandetto — APPROVED** [review]

**Review by bimboterminator1 — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/68415

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1244812

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1244813
